### PR TITLE
Fixes #31395 - Add CV Deb filter for Pulp3

### PR DESCRIPTION
**Pull request by Manisha15**
This PR adds the Content View Debian filter for Pulp3 without dependency_solving and version compare feature.

### CODE REVIEW

**Comment by quba42** [comment]
Note: This is dependent on pulp_deb features available starting with 2.18.0.

I have opened a PR for adding this pulp_deb version to the pulpcore-packaging  3.17 repo here: https://github.com/theforeman/pulpcore-packaging/pull/461

This will also require an up to date pulp-deb-client ruby gem to access the new API endpoint. If needed, I can add a PR for that to foreman-packaging.

**Comment by sjha4** [comment]
We have a repository types endpoint: katello/api/v2/repositories/repository_types which we use in the new CV UI - Repository tab to show types in the Repository Type menu. Would it to possible to leverage that to turn deb options on/off based on results of the API..
katello/webpack/scenes/ContentViews/Details/Repositories/ContentViewRepositories.js
The getRepositoryTypes() action returns enabled repository types and we would likely need that.

Unrelated to this but another area that you might want to look at is the sub-tabs in the version details on new CV UI. We'd probably want to display deb content inside a version to the users especially with filters that can filter out deb content from repos in versions.

**Comment by Manisha15** [comment]
> Looks good overall. I will try to spend some time this week to see if we can leverage the getRepositoryTypes() action to determine if deb is enabled on a box and get back to you with suggestions.

Thanks.

**Comment by Manisha15** [comment]
> @Manisha15 : I am not sure where the on/off switch for Deb is. Can you point me to it?
> 
> Also, I have a separate PR with your commit and a commit with suggested changes on top of your commit to use enabled repo types: [11d8691](https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908)
> 
> Let me know how that looks and feel free to add the commit to this PR if it looks good to you.

I am not sure if there is specific switch to on/off the repository. I think if repository exists in Products then it should be visible.

**Comment by sjha4** [comment]
> > @Manisha15 : I am not sure where the on/off switch for Deb is. Can you point me to it?
> > Also, I have a separate PR with your commit and a commit with suggested changes on top of your commit to use enabled repo types: [11d8691](https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908)
> > Let me know how that looks and feel free to add the commit to this PR if it looks good to you.
> 
> I am not sure if there is specific switch to on/off the repository. I think if repository exists in Products then it should be visible.

Ya..that will work but not in the case of deb type filters. The commit I added in the comment above should fetch the list of enabled repo types and turn that on and off. Feel free to test it and see if it makes sense to you.

**Comment by Manisha15** [comment]
> > > @Manisha15 : I am not sure where the on/off switch for Deb is. Can you point me to it?
> > > Also, I have a separate PR with your commit and a commit with suggested changes on top of your commit to use enabled repo types: [11d8691](https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908)
> > > Let me know how that looks and feel free to add the commit to this PR if it looks good to you.
> > 
> > 
> > I am not sure if there is specific switch to on/off the repository. I think if repository exists in Products then it should be visible.
> 
> Ya..that will work but not in the case of deb type filters. The commit I added in the comment above should fetch the list of enabled repo types and turn that on and off. Feel free to test it and see if it makes sense to you.

I tested the the deb package filtering with your commit and I was still able to see debian repositories and filter the debian packages. Is this the desired functioning or it should have disabled debian repositories and filtering?

**Comment by sjha4** [comment]
That is the desired functionality for boxes with debian content type enabled.

**Comment by Manisha15** [comment]
> That is the desired functionality for boxes with debian content type enabled.

Alright, which means debian repositories are not affected by enabling and disabling of repositories.

**Comment by sjha4** [comment]
@Manisha15 This should be good once the changes here: https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908 are imported into this PR. Let me know if you have any questions/concerns here.

**Comment by Manisha15** [comment]
> @Manisha15 This should be good once the changes here: [11d8691](https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908) are imported into this PR. Let me know if you have any questions/concerns here.

I have imported https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908 to the PR. Since PR requires single commit so I merged both the commits into one but if its accepted to have two commits in PR then I can change it accordingly.

**Comment by Manisha15** [comment]
The webpack test is failing with [11d8691](https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908) , @sjha4 do you know how to fix it?

**Comment by sjha4** [comment]
> The webpack test is failing with [11d8691](https://github.com/Katello/katello/commit/11d8691af1dbd8a58ce45cb6e29ba2c4df877908) , @sjha4 do you know how to fix it?

Yes..We added a new API call for the filter create form. Let me add the nocks in the commit for the test.

**Comment by sjha4** [comment]
Test fixes here: https://github.com/sjha4/katello/commit/02aab3633fd17524a9c403b7893b8f3bd032cf77

Feel free to include if it looks good to you.

File: app/models/katello/content_view_deb_filter.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    def generate_clauses(repo)
+      package_filenames = []
+
+      self.deb_rules.each do |rule|
+        package_filenames.concat(query_debs(repo, rule))
+      end
+
+      if self.original_packages
+        package_filenames.concat(repo.packages_without_errata.map(&:filename))
```
Comment (sjha4) [inline app/models/katello/content_view_deb_filter.rb:17]: Does this method make sense for Deb? Is it being used anywhere?

File: app/models/katello/content_view_deb_filter_rule.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    def ensure_unique_attributes
+      other = self.class.where(:name => self.name,
+                               :version => self.version,
+                               :content_view_filter_id => self.content_view_filter_id,
+                               :min_version => self.min_version,
+                               :max_version => self.max_version)
+      other = other.where.not(:id => self.id) if self.id
+      if other.exists?
+        errors.add(:base, "This package filter rule already exists.")
+      end
```
Comment (sjha4) [inline app/models/katello/content_view_deb_filter_rule.rb:23]: Wording..

File: app/models/katello/content_view_deb_filter_rule.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    validates :name, :presence => true
+    validate :ensure_unique_attributes
+    validates_with Validators::ContentViewFilterVersionValidator
+
+    def ensure_unique_attributes
+      other = self.class.where(:name => self.name,
+                               :version => self.version,
+                               :content_view_filter_id => self.content_view_filter_id,
+                               :min_version => self.min_version,
+                               :max_version => self.max_version)
```
Comment (sjha4) [inline app/models/katello/content_view_deb_filter_rule.rb:19]: These fields don't apply to deb.

File: app/services/katello/pulp3/repository/apt.rb [hunk 4]
```diff
@@ -4,6 +4,9 @@ module Katello
   module Pulp3
     class Repository
       class Apt < ::Katello::Pulp3::Repository
+        include Katello::Util::PulpcoreContentFilters
```
Comment (sjha4) [inline app/services/katello/pulp3/repository/apt.rb:7]: Not necessary.

Comment (Manisha15) [reply app/services/katello/pulp3/repository/apt.rb]: Removed it.

File: app/models/katello/content_view_deb_filter_rule.rb (hunk 3 above)
Comment (Manisha15) [reply app/models/katello/content_view_deb_filter_rule.rb]: Changed the fields.

File: app/models/katello/content_view_deb_filter.rb (hunk 1 above)
Comment (Manisha15) [reply app/models/katello/content_view_deb_filter.rb]: Yeah it's not needed here as we deal with errata separately for debian so I removed this condition as well as function definition.

File: app/models/katello/content_view_deb_filter_rule.rb (hunk 2 above)
Comment (Manisha15) [reply app/models/katello/content_view_deb_filter_rule.rb:21]: Changed the wording.

**Comment by Manisha15** [comment]
I added all the changes but some react test is failing in jenkins which cannot be replicated on my local machine.

**Comment by sjha4** [comment]
> @sjha4 I added all the changes but some react test is failing in jenkins which cannot be replicated on my local machine.

@Manisha15 That is a valid error based on a recent PR we merged to master. I'll push the tests to a commit..

**Comment by sjha4** [comment]
Rubocop failures are relevant.

**Comment by Manisha15** [comment]
> Rubocop failures are relevant.

Fixed them.

**Comment by sjha4** [comment]
For the failing test, could you add:

```
const repoTypesResponse = [{ name: 'deb' }, { name: 'docker' }, { name: 'file' }, { name: 'ostree' }, { name: 'yum' }];
const repoTypeScope = nockInstance
  .get(api.getApiUrl('/repositories/repository_types'))
  .query(true)
  .reply(200, repoTypesResponse);
```
to the start of test# 'Shows call-to-action button when there are no filters' in katello/webpack/scenes/ContentViews/Details/Filters/\__tests\__/contentViewFilters.test.js

You'll also want to check that the call went out with `assertNockRequest(repoTypeScope);` towards the end of the same method. You can add that right before `assertNockRequest(scope, done);`

**Comment by Manisha15** [comment]
> For the failing test, could you add:
> 
> ```
> const repoTypesResponse = [{ name: 'deb' }, { name: 'docker' }, { name: 'file' }, { name: 'ostree' }, { name: 'yum' }];
> const repoTypeScope = nockInstance
>   .get(api.getApiUrl('/repositories/repository_types'))
>   .query(true)
>   .reply(200, repoTypesResponse);
> ```
> 
> to the start of test# 'Shows call-to-action button when there are no filters' in katello/webpack/scenes/ContentViews/Details/Filters/__tests__/contentViewFilters.test.js
> 
> You'll also want to check that the call went out with `assertNockRequest(repoTypeScope);` towards the end of the same method. You can add that right before `assertNockRequest(scope, done);`

I have added your suggestion to failed test, making react tests green. Now only some katello tests are failing but they are not related to this PR.

**Comment by sjha4** [comment]
Ya, ignore those 17 Psych related failures. There's a fix waiting to be merged in foreman that should resolve those.

**Comment by sjha4** [comment]
[test katello]

**Review by sjha4 — APPROVED** [review]
Looks good. Tests are green. Ack 👍🏼

**Comment by sjha4** [comment]
Merging. Thanks @Manisha15 !

**Pull request merged**